### DISCUSSION TITLE
Why does Servo need to use ROUTER ？

### DISCUSSION THREAD

**Discussion by boluochoufeng** [servo/servo#34750] [Q&A]
Hi, I'm here again😂. Why does Servo need to use ROUTER when fetching network requests? I looked at the source code of ipc-channel, and ROUTER seems to only serve as a message forwarding mechanism (not sure). Does it have any other functions? Why is forwarding necessary? Thanks for your answers!

**Comment by jdm** [comment]
The ipc-channel router exists because Servo uses a combination of IPC channels and non-IPC channels. There are many event loops that use the `select` primitive to watch for incoming messages over multiple channels, but it is not possible to mix these channel types. That means that we need a background thread that watches for IPC messages and forwards them to a corresponding non-IPC channel. The router thread performs this service, and the ipc-channel's ROUTER object is the interface to that thread.
